### ISSUE TITLE
Network policies

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#16068]
### Feature request

At present, a Materialize region is available on the public internet without any network-layer access control. We perform access control only at the application layer, via TLS encryption and username/password authentication.

Enterprise users want the ability to specify access control at the network layer via network policies. In practice, this looks like an allowlist of IP CIDR blocks; only connections from those IPs will be permitted. In its simplest form, this could be as simple as a system configuration parameter that specifies the allowed CIDR blocks:

```
SET allowed_ips = ['1.2.3.4/32', '2.3.4.0/24'];
```

We should also look to Snowflake for inspiration. Snowflake permits multiple named network policies, which are manipulated with DDL like [`CREATE NETWORK POLICY`](https://docs.snowflake.com/en/sql-reference/sql/create-network-policy.html) and [`DROP NETWORK POLICY`](https://docs.snowflake.com/en/sql-reference/sql/drop-network-policy.html).

**Comment by ggnall** [comment]
Good example of this ux from clickhouse https://www.loom.com/share/9c6a8d0b27d64fa19e7677625ce94eb1?highlightComment=9362969&t=49

**Comment by benesch** [comment]
> * Is it useful to be able to create multiple named network polices that can be dropped independently?

I think I've figured out the answer to this question. Network policies can be attached to individual users or individual security integrations. See, for example: https://docs.snowflake.com/en/user-guide/network-policies.html#activating-network-policies-for-individual-users. So even though the account as a whole can only have one network policy, individual users may have more restrictive (or more permissive!) network policies.

**Comment by jnoynaert** [comment]
> > * Is it useful to be able to create multiple named network polices that can be dropped independently?
> 
> I think I've figured out the answer to this question. Network policies can be attached to individual users or individual security integrations. See, for example: https://docs.snowflake.com/en/user-guide/network-policies.html#activating-network-policies-for-individual-users. So even though the account as a whole can only have one network policy, individual users may have more restrictive (or more permissive!) network policies.

Being able to scope this to compute (clusters) or data (databases) might be useful as well, further in the future.

**Comment by benesch** [comment]
> Being able to scope this to compute (clusters) or data (databases) might be useful as well, further in the future.

I think that potentially falls out naturally from per-user network policies and role-based access control (#677)! You'd grant access to a specific cluster and/or database to a role, attach a network policy to that role, and then all users with that role would be subject to that network policy.

**Comment by sjwiesman** [comment]
Was recently requested by a customer

**Comment by benesch** [comment]
@jubrad wrote up some thoughts on this on Slack: https://materializeinc.slack.com/archives/C05T1LXH3J5/p1700236298558559

After further thought, I do think we want to treat network policies as named objects. The reason is webhooks. In an environment that's using network policies, we almost _certainly_ want each webhook to be using its own network policy that's different from the network policies that SQL/HTTP connections use. Also, as mentioned in the last comment, we likely eventually want the ability to have different network policies per user.

So, here's a slightly more fleshed out design for the network policy syntax:

```sql
CREATE NETWORK POLICY <name> (
    ALLOW IP ADDRESSES = ('<CIRD 1>', '<CIDR 2>', ...),
    DENY IP ADDRESSES = ('<CIRD 1>', '<CIDR 2>', ...)
)
```

Like clusters, network policies exist outside of databases or schemas in a global namespace. To create a network policy, you need the `CREATENETWORKPOLICY` system privilege. To _use_ a network policy, you need the `USAGE` privilege on the policy. There are three places a network policy can be used:

```
-- Set the default network policy if no more specific network policy applies. 
ALTER SYSTEM SET network_policy = <name>;

-- Set the network policy for a user.
ALTER ROLE <role> SET network_policy = <name>;

-- Set the network policy for a webhook source.
CREATE SOURCE <webhook> (NETWORK POLICY = <name>);
```

Network policies do not inherit. The most specific network policy wins. E.g., if a webhook source has a network policy explicitly configured, only that network policy applies. Otherwise, the account's default network policy applies. If no such network policy exists, then all IPs are allowed.

**Comment by jubrad** [comment]
This seems great!

> Network policies do not inherit. The most specific network policy wins

Do all policies, allow and deny, get placed in order by specificity or allows grouped first then denies? Either way, having a query /show command that displays the rules broken down by CIDRs in the order they'll be applied would be a nice touch.

**Comment by benesch** [comment]
> Do all policies, allow and deny, get placed in order by specificity or allows grouped first then denies?

Yeah, I propose we follow Snowflake here. Deny by default, followed by all allow rules, followed by all deny rules.

We also need to set some limits. Maybe no more than 32 entries in each of `ALLOW IPS` and `DENY IPS`?

> Either way, having a query /show command that displays the rules broken down by CIDRs in the order they'll be applied would be a nice touch.

Agreed! Possibly we want the command to expand out the allow CIDRS and subtract out the deny CIDRs, to show you exactly the CIDRs that are allowed?

**Comment by jubrad** [comment]
(allow list - deny list) with deny fall through sounds good.

32 is probably a good number, at least to start. I think having too many might give someone the impression they can do broad allows and and narrow down with deny, which is definitely not the goal.

A command to show the exact allowed CIDRs would be great.
This is definitely getting ahead of things, but I could see this being useful, along with a historical view of sessions and the ips they  occurred on, to audit access after discovering a flawed policy (ex: a deny cidr missing). Although, for that, we'd also need some ip/cidr functions to make it manageable in sql.

**Comment by benesch** [comment]
> This is definitely getting ahead of things, but I could see this being useful, along with a historical view of sessions and the ips they occurred on, to audit access after discovering a flawed policy (ex: a deny cidr missing). Although, for that, we'd also need some ip/cidr functions to make it manageable in sql.

Useful indeed! PostgreSQL has some IP related functions and types that we should import. Also that historical view of sessions already exists! `mz_session_history`. Should make sure to add the client IP to that table...

**Comment by benesch** [comment]
Added a note to the issue description about supporting network policies for service users (#20894).

Slack discussion: https://materializeinc.slack.com/archives/C06GZ7GBKB5/p1718974403636149?thread_ts=1718124599.991419&cid=C06GZ7GBKB5 (cc @morsapaes)
